### PR TITLE
Reapply "[DAGCombiner] Add support for scalarising extracts of a vector setcc  (#117566)"

### PR DESCRIPTION
**Pull request by david-arm**
[Reverts d57892a2a153ab71a796f07e39d939eae6910c21]

For IR like this:

%icmp = icmp ult <4 x i32> %a, splat (i32 5)
%res = extractelement <4 x i1> %icmp, i32 1

where there is only one use of %icmp we can take a similar approach
to what we already do for binary ops such add, sub, etc. and convert
this into

%ext = extractelement <4 x i32> %a, i32 1
%res = icmp ult i32 %ext, 5

For AArch64 targets at least the scalar boolean result will almost
certainly need to be in a GPR anyway, since it will probably be
used by branches for control flow. I've tried to reuse existing code
in scalarizeExtractedBinop to also work for setcc.

NOTE: The optimisations don't apply for tests such as
extract_icmp_v4i32_splat_rhs in the file

CodeGen/AArch64/extract-vector-cmp.ll

because scalarizeExtractedBinOp only works if one of the input
operands is a constant.

### CODE REVIEW

**Comment by david-arm** [comment]
Yet another attempt to reland https://github.com/llvm/llvm-project/pull/117566. The last post-commit failure exposed an existing issue where we were missing a `ExpandIntRes_SETCC` to handle i128 result types. I've added a new test for this, but it can only be exposed once the new DAG combine kicks in.

**Review by arsenm — COMMENTED** [review]
I thought this already landed?

**Comment by paulwalker-arm** [comment]
> I thought this already landed?

The patch was reverted because it exposed a bug within legalisation (expansion) of i128 based setcc operations.

**Comment by arsenm** [comment]
> > I thought this already landed?
> 
> The patch was reverted because it exposed a bug within legalisation (expansion) of i128 based setcc operations.

Should have the proper git revert with the reapply in the message

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp [hunk 1]
```diff
@@ -3316,6 +3317,22 @@ static std::pair<ISD::CondCode, ISD::NodeType> getExpandedMinMaxOps(int Op) {
...
+  SDValue LHS = N->getOperand(0);
+  SDValue RHS = N->getOperand(1);
+  EVT NewVT = getSetCCResultType(LHS.getValueType());
+
+  // Taking the same approach as ScalarizeVecRes_SETCC
+  SDValue Res = DAG.getNode(ISD::SETCC, DL, NewVT, LHS, RHS, N->getOperand(2));
+
+  ISD::NodeType ExtendCode =
+      TargetLowering::getExtendForContent(TLI.getBooleanContents(NewVT));
+  Res = DAG.getExtOrTrunc(Res, DL, N->getValueType(0), ExtendCode);
```
Comment (paulwalker-arm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:3332]:
```suggestion
  Res = DAG.getBoolExtOrTrunc(Res, DL, N->getValueType(0), NewVT);
```

**Review by paulwalker-arm — APPROVED** [review]
Hopefully third time's the charm.

**Pull request merged**